### PR TITLE
Carousel: Fix autoplay and enable callback on autoplay index change

### PR DESCRIPTION
**Pull request by Mitch-At-Work**
## Previous Behavior
No autoplay callback making controlled scenarios not usable with autoplay.
Autoplay would start with 'stopOnInteraction' based on its autoplay value, this caused three issues:
- stopOnInteraction: true would stop autoplay and not resume it after interactions
- stopOnInteraction: false would restart autoplay functionality even if autoplay set to false
- stopOnInteraction was not updated unless a re-init or recreation of Embla occurred, prevented by ref value not driving updates

## New Behavior
Autoplay callback enables controlled autoplay callbacks
Autoplay will now be attached always, but conditionally enabled by stop/start calls, removing need for a reinitialize of carousel
Autoplay logic can be paused/resumed without overriding play button state for temporary interaction pauses like focus or mouse hover

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -218,9 +218,9 @@
     "doctrine": "3.0.0",
```
Comment (fabricteam) [inline package.json:218]: ## 🕵🏾‍♀️ [<b data-uid='vrt-gist'></b> visual regressions to review in the fluentuiv8 Visual Regression Report](https://aka.ms/visual-changes-review?organization=uifabric&project=fabricpublic&prId=33291&commitId=f645db6e16288044a28260bb2b633b16386eb88e&env=prod&clientType=FLUENTUI)


  <details><summary><b>Callout</b> 1 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Callout.Top left edge.chromium.png|2307| Changed|

</details>


  <details><summary><b>react-charting-LineChart</b> 1 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|react-charting-LineChart.Gaps.chromium.png|1| Changed|

</details>




<div id="vrtCommentfluent-ui_VRT_Pipeline_v8"/>

**Comment by fabricteam** [comment]
## 🕵 fluentui-web-components-v3 No visual regressions between this PR and main
<div id="vrtCommentfluent-ui_VRT_Pipeline_web-components"/>

File: package.json (hunk 1 above)
Comment (fabricteam) [inline package.json:218]: ## 🕵🏾‍♀️ [<b data-uid='vrt-gist'></b> visual regressions to review in the fluentuiv9 Visual Regression Report](https://aka.ms/visual-changes-review?organization=uifabric&project=fabricpublic&prId=33291&commitId=f645db6e16288044a28260bb2b633b16386eb88e&env=prod&clientType=FLUENTUI)


  <details><summary><b>Avatar Converged</b> 2 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Avatar Converged.badgeMask.chromium.png|3| Changed|
|Avatar Converged.Badge Mask RTL.chromium.png|1| Changed|

</details>


  <details><summary><b>Drawer</b> 1 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Drawer.Full Overlay High Contrast.chromium.png|6795| Changed|

</details>




<div id="vrtCommentfluent-ui_VRT_Pipeline_v9"/>

Comment (brandonthomas) [reply package.json:218]: unrelated?

Comment (brandonthomas) [reply package.json:218]: unrelated?

File: packages/react-components/react-carousel/library/src/components/useEmblaCarousel.ts [hunk 2]
```diff
@@ -114,6 +114,53 @@ export function useEmblaCarousel(
...
+        ...DEFAULT_EMBLA_OPTIONS,
+        ...emblaOptions.current,
+      },
+      plugins,
+    );
+  }, [align, containScroll, direction, getPlugins, loop, slidesToScroll, watchDrag]);
+
+  React.useEffect(() => {
+    reinitializeCarousel();
+  }, [reinitializeCarousel]);
```
Comment (layershifter) [inline packages/react-components/react-carousel/library/src/components/useEmblaCarousel.ts:141]: I don't see `reinitializeCarousel` being used anywhere, why this change is needed? Can we revert it?

File: packages/react-components/react-carousel/library/src/components/useEmblaCarousel.ts [hunk 3]
```diff
@@ -142,22 +189,27 @@ export function useEmblaCarousel(
...
+  }, [setActiveIndex]);
+
+  const handleAutoplayIndexChange = useEventCallback(() => {
+    handleIndexChange();
+    const newIndex = emblaApi.current?.selectedScrollSnap() ?? 0;
+    onAutoplayIndexChange?.(null, { event: null, type: 'autoplay', index: newIndex });
+  });
+
   const viewportRef: React.RefObject<HTMLDivElement> = React.useRef(null);
+  const currentElementRef = React.useRef<HTMLDivElement | null>();
```
Comment (layershifter) [inline packages/react-components/react-carousel/library/src/components/useEmblaCarousel.ts:211]: We don't use this ref anywhere, correct? Let's restore `currentElement`, so it won't leak the scope of `useMemo()`.

File: packages/react-components/react-carousel/library/src/components/useEmblaCarousel.ts [hunk 4]
```diff
@@ -172,21 +224,23 @@ export function useEmblaCarousel(
...
           emblaApi.current?.off('select', handleIndexChange);
           emblaApi.current?.off('reInit', handleReinit);
+          emblaApi.current?.off('autoplay:select', handleAutoplayIndexChange);
           emblaApi.current?.destroy();
         }
 
         // Use direct viewport if available, else fallback to container (includes Carousel controls).
         const wrapperElement = viewportRef.current ?? newElement;
+        currentElementRef.current = wrapperElement;
         if (wrapperElement) {
```
Comment (layershifter) [inline packages/react-components/react-carousel/library/src/components/useEmblaCarousel.ts:243]:
```suggestion
        currentElement = viewportRef.current ?? newElement;

        if (currentElement) {
          emblaApi.current = EmblaCarousel(
            currentElement,
```

File: packages/react-components/react-carousel/library/src/components/useEmblaCarousel.ts [hunk 5]
```diff
@@ -67,36 +78,25 @@ export function useEmblaCarousel(
   });
 
   const emblaApi = React.useRef<EmblaCarouselType | null>(null);
-  const autoplayRef = React.useRef<boolean>(false);
+  const autoplay = React.useRef<boolean>(false);
```
Comment (layershifter) [inline packages/react-components/react-carousel/library/src/components/useEmblaCarousel.ts:81]: nit, let's keep `autoplayRef` as it makes its clearer

File: packages/react-components/react-carousel/library/src/components/useEmblaCarousel.ts [hunk 6]
```diff
@@ -142,22 +189,27 @@ export function useEmblaCarousel(
     }
   });
 
+  const handleIndexChange = React.useCallback(() => {
```
Comment (layershifter) [inline packages/react-components/react-carousel/library/src/components/useEmblaCarousel.ts:192]:
```
type EmblaEventHandler = Parameters<EmblaCarouselType['on']>[1];

// ...
```

and then we can do:

```
  const handleIndexChange: EmblaEventHandler = useEventCallback((_, eventType) => {
    const newIndex = emblaApi.current?.selectedScrollSnap() ?? 0;
    const slides = emblaApi.current?.slideNodes();
    const actualIndex = emblaApi.current?.internalEngine().slideRegistry[newIndex][0] ?? 0;

    // We set the active or first index of group on-screen as the selected tabster index
    slides?.forEach((slide, slideIndex) => {
      setTabsterDefault(slide, slideIndex === actualIndex);
    });
    setActiveIndex(newIndex);

    if (eventType === 'autoplay:select') {
      onAutoplayIndexChange?.(null, { event: null, type: 'autoplay', index: newIndex });
    }
  });
```

🐱

File: packages/react-components/react-carousel/library/etc/react-carousel.api.md [hunk 7]
```diff
@@ -209,6 +209,7 @@ export type CarouselProps = ComponentProps<CarouselSlots> & {
     align?: 'center' | 'start' | 'end';
     activeIndex?: number;
     onActiveIndexChange?: EventHandler<CarouselIndexChangeData>;
+    onAutoplayIndexChange?: (ev: null, data: CarouselAutoplayIndexChangeData) => void;
```
Comment (layershifter) [inline packages/react-components/react-carousel/library/etc/react-carousel.api.md:212]: While it sucks, probably it would be easier to pass just `new Event()` instead of having two separate callbacks, so consumers will have a single source of truth

File: packages/react-components/react-carousel/library/src/components/useEmblaCarousel.ts (hunk 6 above)
Comment (Mitch-At-Work) [reply packages/react-components/react-carousel/library/src/components/useEmblaCarousel.ts:192]: Sadly the eventType is generic so we only see a 'select' event on autoplay index change

**Review by dmytrokirpa — APPROVED** [review]
Everything's looking good on the v-build side!

File: packages/react-components/react-carousel/library/src/components/useEmblaCarousel.ts [hunk 8]
```diff
@@ -114,6 +116,27 @@ export function useEmblaCarousel(
...
+        // Autoplay should only enable in the case where underlying state is true, temporary should not override
+        emblaApi.current?.plugins().autoplay?.play();
+        // Reset after play to ensure timing and any focus/mouse pause state is reset.
+        resetAutoplay();
+      } else if (!_autoplay) {
+        emblaApi.current?.plugins().autoplay?.stop();
+      }
+    },
+    [resetAutoplay],
+  );
```
Comment (layershifter) [inline packages/react-components/react-carousel/library/src/components/useEmblaCarousel.ts:138]:
```suggestion
  const enableAutoplay = React.useCallback(
    (autoplay: boolean, temporary?: boolean) => {
      if (!temporary) {
        autoplayRef.current = autoplay;
      }

      if (autoplay && autoplayRef.current) {
        // Autoplay should only enable in the case where underlying state is true, temporary should not override
        emblaApi.current?.plugins().autoplay?.play();
        // Reset after play to ensure timing and any focus/mouse pause state is reset.
        resetAutoplay();
      } else if (!autoplay) {
        emblaApi.current?.plugins().autoplay?.stop();
      }
    },
    [resetAutoplay],
  );
```

nit: we don't need `_` after the latest changes 🐱

File: packages/react-components/react-carousel/library/src/components/useEmblaCarousel.ts [hunk 9]
```diff
@@ -142,22 +165,27 @@ export function useEmblaCarousel(
...
+
+    // We set the active or first index of group on-screen as the selected tabster index
+    slides?.forEach((slide, slideIndex) => {
+      setTabsterDefault(slide, slideIndex === actualIndex);
+    });
+    setActiveIndex(newIndex);
+
+    if (eventType === 'autoplay:select') {
+      const _event = new Event('autoplay');
+      onAutoplayIndexChange?.(_event, { event: _event, type: 'autoplay', index: newIndex });
```
Comment (layershifter) [inline packages/react-components/react-carousel/library/src/components/useEmblaCarousel.ts:181]:
```suggestion
      // There is no event object for autoplay as there is no a DOM event
      const noopEvent = new Event('autoplay');

      onAutoplayIndexChange?.(noopEvent, { event: noopEvent, type: 'autoplay', index: newIndex });
```

**Review by layershifter — APPROVED** [review]

File: change/@fluentui-react-carousel-806c25e9-e23b-4c65-b072-007f894a236a.json [hunk 10]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (fabricteam) [inline change/@fluentui-react-carousel-806c25e9-e23b-4c65-b072-007f894a236a.json:1]: ## 🕵🏾‍♀️ [<b data-uid='vrt-gist'></b> visual regressions to review in the fluentui-web-components-v3 Visual Regression Report](https://aka.ms/visual-changes-review?organization=uifabric&project=fabricpublic&prId=33291&commitId=0289c7f74239477975bca94fa158f29e27b9bb66&env=prod&clientType=FLUENTUI)


  <details><summary><b>Avatar</b> 1 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Avatar.Image.chromium.png|248| Changed|

</details>




<div id="vrtCommentfluent-ui_VRT_Pipeline_web-components"/>
